### PR TITLE
Fix faulty Spiro handling in SplinePointListSet by removing it, replacing it with SplinePointListCopy

### LINKED ISSUE #4037: Crash while scaling spline (perhaps Spiro related)
2019-11-28-193852_1978x1121_scrot

1. Open [53.sfd](https://github.com/fontforge/fontforge/files/3901432/53.zip)
2. Open glyph `five`
3. Verify Spiro mode is enabled
4. Select all points in the digit five by clicking any of its points
4. Scale tool
5. Start scale at corner point highlighted in image
6. Scale size to match size of smaller three
7. You should get the following, in my experience highly reliable, crash:

```gdb
Program received signal SIGSEGV, Segmentation fault.
0x0000155555104f8c in SplinePointListSet (tobase=0x555555dc8840, 
    frombase=0x555555f4f760) at ../fontforge/splineutil2.c:3648
3648		for ( tsp = tobase->first, fsp=frombase->first; tsp!=first; tsp=tsp->next->to, fsp=fsp->next->to ) {
(gdb) bt
#0  0x0000155555104f8c in SplinePointListSet (tobase=0x555555dc8840, frombase=0x555555f4f760) at ../fontforge/splineutil2.c:3648
#1  0x0000155554e454e5 in _CVRestoreTOriginalState (cv=0x15554ed48010, p=0x15554ed4b3d0) at ../fontforge/cvundoes.c:1080
#2  0x000055555567e51f in CVRestoreTOriginalState (cv=0x15554ed48010) at ../fontforgeexe/cvpointer.c:549
#3  0x00005555556a5957 in CVMouseMoveTransform (cv=0x15554ed48010) at ../fontforgeexe/cvtranstools.c:44
#4  0x0000555555603c5f in CVMouseMove (cv=0x15554ed48010, event=0x7fffffffd130) at ../fontforgeexe/charview.c:5422
#5  0x0000555555604fe8 in v_e_h (gw=0x555555f048f0, event=0x7fffffffd130) at ../fontforgeexe/charview.c:5732
#6  0x000055555584e5bc in _GWidget_Container_eh (gw=0x555555f048f0, event=0x7fffffffd130) at ../gdraw/gcontainer.c:405
#7  0x0000555555862885 in _GGDKDraw_CallEHChecked (gw=0x555555f048f0, event=0x7fffffffd130, eh=0x55555584d91c <_GWidget_Container_eh>) at ../gdraw/ggdkdraw.c:334
#8  0x00005555558652a8 in _GGDKDraw_DispatchEvent (event=0x555555f1a320, data=0x555555be1a40) at ../gdraw/ggdkdraw.c:1237
#9  0x0000155554195954 in  () at /usr/lib/libgdk-3.so.0
#10 0x0000155554142904 in  () at /usr/lib/libgdk-3.so.0
#11 0x0000155553d1639e in g_main_context_dispatch () at /usr/lib/libglib-2.0.so.0
#12 0x0000155553d181b1 in  () at /usr/lib/libglib-2.0.so.0
#13 0x0000155553d181f1 in g_main_context_iteration () at /usr/lib/libglib-2.0.so.0
#14 0x0000555555868052 in GGDKDrawEventLoop (gdisp=0x555555be1a40) at ../gdraw/ggdkdraw.c:2204
#15 0x0000555555853ada in GDrawEventLoop (gdisp=0x555555be1a40) at ../gdraw/gdraw.c:789
#16 0x0000555555817e7b in fontforge_main (argc=2, argv=0x7fffffffe568) at ../fontforgeexe/startui.c:1415
#17 0x00005555555aab89 in main (argc=2, argv=0x7fffffffe568) at ../fontforgeexe/main.c:33
```

### PR DESCRIPTION
**Pull request by ctrlcctrlv**
Sorry for the duplicate issue and all the noise. 

This closes #4037. This is a pared down version of the previous version of this PR, #4038, which I found to cause bugs elsewhere.

### CODE REVIEW

File: fontforge/splineutil2.c [hunk 1]
```diff
@@ -3654,12 +3654,12 @@ void SplinePointListSet(SplinePointList *tobase, SplinePointList *frombase) {
 	    tsp->noprevcp = fsp->noprevcp;
 	    tsp->nonextcp = fsp->nonextcp;
 	    if ( first==NULL ) first = tsp;
-	    if ( tsp->next==NULL )
+	    if ( tsp->next==NULL || fsp->next==NULL )
```
Comment (ctrlcctrlv) [inline fontforge/splineutil2.c]: The problem @frank-trampe as I understand it is that the splines get out of sync due to spiros, and soon end up different lengths, an impossible situation with Beziers but perfectly possible with Spiros as the Spiro&rarr;Bezier code guarantees nothing. So this extra sanity check just makes sure we actually have more points to copy from.

File: fontforge/splineutil2.c [hunk 2]
```diff
@@ -3654,12 +3654,12 @@ void SplinePointListSet(SplinePointList *tobase, SplinePointList *frombase) {
...
 	break;
 	}
 	firstsp = NULL;
 	for ( tspline = tobase->first->next, fspline=frombase->first->next;
-		tspline!=NULL && tspline!=firstsp;
+		tspline!=NULL && frombase->spiro_cnt==0 && tspline!=firstsp;
 		tspline=tspline->to->next, fspline=fspline->to->next ) {
```
Comment (ctrlcctrlv) [inline fontforge/splineutil2.c]: This prevents the whole block from being executed @frank-trampe if there are Spiros as the Spiro copy below does the job. Without it we get crash:

```
Program received signal SIGSEGV, Segmentation fault.
0x0000155555105095 in SplinePointListSet (tobase=0x555555d5ac80, frombase=0x555555dfc230) at ../fontforge/splineutil2.c:3664
3664		   tspline->splines[0] = fspline->splines[0];
```

`fspline` is NULL!

**Comment by skef** [comment]
@frank-trampe This function is used in exactly two places, both within the same cvundoes.c call. In this day and age I would call it an example of over-optimization against memory allocation. Can we please tell @ctrlcctrlv to use `SplinePointListCopy()` instead, and delete this? 

@ctrlcctrlv Would you be willing to test with the caller doing
```
    SplinePointListFree(cv->layerheads[cv->drawmode]->splines);
    cv->layerheads[cv->drawmode]->splines = SplinePointListCopy(undo->u.state.splines);
```
in the first instance with the equivalent change in the second?

**Comment by ctrlcctrlv** [comment]
Works great @skef, good idea!

**Comment by skef** [comment]
This comes from my suggestion so I'm not going to officially approve it, but unofficially I very much approve *of* it. 

Looking at the code a bit more, the idea seems to have been that there are some contexts where you know the operation preserves the point count, such as a transformation, so why not just reuse the list of points? And one answer is: because you might be doing the operation on Spiro points for which the cubic Bezier "translation" doesn't have that property. 

So it's possible that this code should be further modified to get rid of the function calls that relate to the original faulty assumption. I think that can be left to the future but I'll let others decide.

**Comment by frank-trampe** [comment]
Fine with me. It's also fine to merge a pull request with two sets of competent eyes on it, regardless of who originated the ideas and the code.

**Comment by frank-trampe** [comment]
We can make further improvements as needed. If this solves Fred's problem, I say we merge it now.

**Comment by skef** [comment]
>  It's also fine to merge a pull request with two sets of competent eyes on it, regardless of who originated the ideas and the code.

Yeah -- I mostly already do that with straightforward bug fixes but I'm more reluctant when a decision involves a tradeoff. In a way I felt too strongly about this one to be comfortable also being the one that signs off, if that make sense.

**Pull request merged**